### PR TITLE
TINY-7705: Fixed table keyboard navigation trying to place the selection inside a noneditable cell

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: TINY-7705

Description of Changes:
* Update the snooker cell navigation logic to all a predicate to be passed so that we can walk over noneditable elements
* Added a new ContentEditable module to sugar to handle checking if a HTML element is editable. Also updated some other places in snooker to use this new shared logic.
* Fixed the table tabbing logic trying to place the selection inside a noneditable cell by using the above 2 changes.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Review by diogo405 — APPROVED** [review]

File: modules/tinymce/src/core/demo/html/full_demo.html [hunk 1]
```diff
@@ -12,7 +12,7 @@ <h1>Tinymce full featured Page</h1>
     <div id="ephox-ui">
       <textarea></textarea>
     </div>
-    <div class="tinymce">Inline demo</div>
+    <div class="tinymce"><table><tbody><tr><td>Inline demo</td><td>Cell 2</td></tr></tbody></table></div>
```
Comment (ashfordneil) [inline modules/tinymce/src/core/demo/html/full_demo.html:15]: Why was this changed?

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/core/demo/html/full_demo.html (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/core/demo/html/full_demo.html:15]: This was not meant to be committed... thanks for the catch!

- [ ] Revert this

File: modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts [hunk 2]
```diff
@@ -1,14 +1,44 @@
...
+): CellLocation => {
+  const forwards = direction > 0;
+  if (!forwards && index <= 0) {
+    return CellLocation.first(all[0]);
+  } else if (forwards && index >= all.length - 1) {
+    return CellLocation.last(all[all.length - 1]);
+  } else {
+    const newIndex = index + direction;
+    const elem = all[newIndex];
+    return isEligible(elem) ? CellLocation.middle(current, elem) : walk(all, current, newIndex, direction, isEligible);
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts:33]: I had considered that, but we should never to ever be looping often here (e.g it's in the realm of 2 or 3 times) so it wasn't really worth the complexities of using a loop.

File: modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts [hunk 3]
```diff
@@ -1,14 +1,44 @@
...
+type IsRootFn = (e: SugarElement<Node>) => boolean;
+
+/*
+ *  Walk until the next eligible cell location is found, or the start/end of the table is found.
+ */
+const walk = (
+  all: SugarElement<HTMLTableCellElement>[],
+  current: SugarElement<HTMLTableCellElement>,
+  index: number,
+  direction: number,
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts:22]: I didn't realise you could do a union for actual numerical values... TIL. I'll change it to that as we want it to be a number so we can easily make the index changes.

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts:22]: Actually we can probably do that with a `const enum`, so will investigate that as well.

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts]: Changed to a constant enum.

File: modules/tinymce/src/core/demo/html/full_demo.html (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/core/demo/html/full_demo.html]: Reverted

File: modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/CellNavigation.ts:38]: I should add this isn't a "hot code path" so saving the <1ms per iteration for the call stack isn't worth considering here. If this was going to be called hundreds of times then it'd be worth looking at.

**Review by ltrouton — APPROVED** [review]

File: modules/sugar/CHANGELOG.md [hunk 4]
```diff
@@ -6,5 +6,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## Unreleased
 
+### Added
+- Added new `ContentEditable` module to determine if a HTML element is content editable.
```
Comment (JamesToohey) [inline modules/sugar/CHANGELOG.md:10]:
```suggestion
- Added new `ContentEditable` module to determine if an HTML element is content editable.
```

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 5]
```diff
@@ -42,8 +42,10 @@ interface ExtractColsDetail {
 
 type CompElm = (e1: SugarElement, e2: SugarElement) => boolean;
 
+// This uses a slight variation to the default `ContentEditable.isEditable` behaviour,
+// as when the element is detached we assume it is editable instead as it'll be a new cell.
```
Comment (JamesToohey) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:46]: Just a slight tweak to avoid the two 'as' in a row - up to you

```suggestion
// This uses a slight variation to the default `ContentEditable.isEditable` behaviour,
// as when the element is detached we assume it is editable because it is a new cell.
```

**Review by JamesToohey — APPROVED** [review]
